enhancement: user workload monitoring (#26)
openshift/enhancements: enhancements/user-monitoring.md

---
title: user-monitoring
authors:
  - "@s-urbaniak"
  - "@brancz"
reviewers:
  - "@sichvoge"
  - "@openshift/openshift-team-monitoring"
approvers:
  - "@sichvoge"
  - "@openshift/openshift-team-monitoring"
creation-date: 2019-09-19
last-updated: 2019-09-19
status: implementable
---

# user-monitoring

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined

> - [ ] Graduation criteria for dev preview, tech preview, GA

Reviewer: We can mark this as well?

Author: hmm, i think we did not specify the actual _criteria_ for GA.

- [ ] User-facing documentation is created in [openshift/docs]

## Summary

> This enhancement describes a new feature for OpenShift cluster observability, namely user monitoring.

Reviewer 2: user workload monitoring?  user namespace monitoring?  we aren't monitoring the users themselves :)

Reviewer 3: user defined monitoring?

Reviewer 4: service monitoring?

Reviewer: Yep user workload monitoring is self descriptive enough. Service monitoring makes me think of k8s services.

Reviewer 3: Maybe "user defined monitoring" ? Generally agree with Christian though.

It is an extension to the existing cluster monitoring stack to enable observability of user namespaces.
Currently, the cluster monitoring stack only reconciles OpenShift namespaces.

## Motivation

Currently there is no supported solution for customers to monitor their business critical applications.
The existing cluster monitoring stack has no support for adding scrape targets in user namespaces
as it is only available for OpenShift components scraping workloads from `openshift-*` namespaces.
Other Red Hat products, outside the OpenShift framework components
don’t have a monitoring solution to integrate their metrics and out-of-the-box insights into their well being.

> Currently it is difficult to achieve metering/chargeback for Red Hat products without a single source.

Reviewer 4: Currently, customers have to spin up their own Prometheus ecosystem parallel to our stack. In some cluster you may end up with multiple different "stacks" (many sources). The goal is to avoid that as it complicates integration with other services like Metering or the OpenShift console, and instead provide a "single source" that they can always rely on.  

guess it refers to having teams/customers to spin up their own prometheus ecosystem which would mean "many sources" that contain critical metrics for, for example, chargeback/showback use cases.

> Currently it is difficult to achieve metering/chargeback for Red Hat products without a single source.

Author: I (hopefully) clarified this via `s/single source/central observability stack`.

### Goals

> Allow deployment and reconcilation scrape targets, recording rules, and alerting rules in user namespaces.

Reviewer 2: i would say deploying a second prometheus is not a goal.  it may be an implementation requirement.

> Allow deployment and reconcilation scrape targets, recording rules, and alerting rules in user namespaces.

Reviewer 5: can agree, removed

> Allow deployment and reconcilation scrape targets, recording rules, and alerting rules in user namespaces.

Reviewer 3: querying cluster metrics is already possible today, and customizations to the stack itself are not going to be possible in the first iteration

> Allow deployment and reconcilation scrape targets, recording rules, and alerting rules in user namespaces.

Reviewer 4: correlation between app and cluster centric metrics is a good goal as well. i heard customers talking about that "capability" before.

> Allow deployment and reconcilation scrape targets, recording rules, and alerting rules in user namespaces.

Author: left out customizations, added correlation bullet point.

> This effectively allows the reconcilation of the following custom resources in user namespaces:

Author: sounds good, replaced :+1:

> - ServiceMonitor (https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#servicemonitor)

Reviewer: Were these meant to be links? e.g. `[](url)`

Author: hmm, i didn't mean them to be links but agree it looks better :-) fixed

- PodMonitor (https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#podmonitor)
- PrometheusRule (https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#prometheusrule)

### Non-Goals

> Out of scope are:

Author: fixed

- Definition of Dashboards.
- Deployment of Prometheus instances in user namespaces.

> - Deployment of Alertmanager instances user namespaces.

Reviewer 6: (nit) instances **in** user namespaces

Author: fixed

## Proposal

![](user-monitoring-deployment.png)

The existing cluster monitoring stack is going to be extended with additional Prometheus (P) servers

> and one additional Prometheus Operator (PO) reconciling user namespaces

Reviewer 2: why a new operator?  can the existing operator not do the additional work?  is there an advantage to having it as a separate operator?  (it's still going to be CVO managed, right?)

as well as the newly added Prometheus servers.

The cluster Prometheus Operator (PO) continues to operate the cluster-level Prometheus.

> We will continue to use a centralized Alertmanager (A) cluster.

Reviewer 2: will this central alertmanager now include alerts for user namespace content?

Reviewer 2: elsewhere it sounds like we'll have an alertmanager per user namespace?

if alertmanager *is* shared, what stops me from defining an alert in terms of someone else's metrics, as a way to get those metric values sent to me?

Reviewer 3: This is an implementation detail that goes deeper than this proposal should go in my opinion, but the Prometheus operator will force specific labels onto metrics and alerts by which a central Alertmanager config can route (in the same way as we force specific labels for query purposes).

Author: added the following clarifying sentence:
> We will continue to use a centralized Alertmanager (A) cluster. The existing cluster will aggregate both user workload alerts as well as cluster alerts. Tenancy will be achieved by forcing specific labels onto metrics and alerts by which the central Alertmanager config can route.

Querying data from a single, multi-tenant interface is done using the Thanos Querier (TQ) component.

> Soft tenancy is continued to be ensured by the existing topology leveraging kube-rbac-proxy [1]

Reviewer 7: "is continued" -> "continues"

> Soft tenancy is continued to be ensured by the existing topology leveraging kube-rbac-proxy [1]

Reviewer 2: can we define soft tenancy?

> Soft tenancy is continued to be ensured by the existing topology leveraging kube-rbac-proxy [1]

Reviewer 3: I think we can just say "tenancy" and explain how we achieve it, which is by enforcing tenancy at the query layer.

> Soft tenancy is continued to be ensured by the existing topology leveraging kube-rbac-proxy [1]

Author: replaced "soft tenancy" with "tenancy"

and prom-label-proxy [2].

Differences between cluster- and user-level is mainly upgrade guarantees.

> As we keep cluster-level Prometheus immutable, we can guarantee a 100% cluster upgrade success

Reviewer 3: There is no such thing as 100%. "The cluster-level Prometheus stack is an integral part of OpenShift and its reliability is important, keeping the stack immutable and fully testable by us allows us to rely on it to a higher degree. As we have no such possibility for user defined configurations and scrape targets, we can't provide the same guarantees."

Author: took over that sentence :+1:

but that rate may be lower for any additional Prometheus server scraping user namespaces.
There are edge cases where an automated upgrade is not possible in the face of user monitoring.
If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions

> we most likely cannot perform an automated migration, as PromQL is too complex to achieve that.

Reviewer 5: I think it's highly unlikely that we break any compatibility with PromQL, also argument " as PromQL is too complex to achieve that." is not very convincing to me (: Can we change to maybe:

 ```suggestion
we most likely cannot perform an automated migration. Such migration without knowledge of user intention behind each query would be exceptionally difficult.
```

Reviewer 3: Generally agreed, but for example we dropped a few PromQL functions in 1.x -> 2.x. A migration would have been impossible, as the functions simply didn't exist anymore and beyond that, modifying user controlled configuration that they probably keep in git repos sounds like a bad idea to begin with.

Reviewer 5: Definitely, but also I think there are way more disruptive changes in those major versions updates (e.g storage format!) then just PromQL. Maybe worth to mention that?

Author: @bwplotka storage format is something that is under supervision of openshift namespaces, not user namespaces, hence under "tight" control. If upstream breaks that, we will most likely need to implement automatic migrations into cluster monitoring operator and/or prometheus operator.

> Instead the "upgradable" condition in the ClusterOperator will be set to false.

Reviewer 5: ```suggestion
Instead, the "upgradable" condition in the ClusterOperator will be set to false.
```

Reviewer 5: Can we follow up with what this means? (newbie here) (:

Reviewer 5: How users will upgrade / what are the consequences?

Reviewer 2: and how would we detect this condition (that there were non-upgradeable queries?)  we have to set the condition before we even update the operator, so how would the old operator know that the new operator can't handle certain content?

Reviewer 3: This is where @bwplotka's above comment does come in. We do know about the changes to PromQL exceptionally well so if there was a breaking change, then we could look at user defined configuration and know if a PromQL statement were to become invalid. As Bartek said, this is highly unlikely but has happened once in the past.

> Routing configurations for alerts to be sent to specific teams need to be communicated/configured by an infrastructure owner.

Reviewer 3: Maybe add "... using the existing Alertmanager stack that is part of cluster-monitoring."

Author: added

> We will look into better mechanisms after GA.

Reviewer 3: "better" is subjective, "extension mechanisms post GA." is appropriate I think

Author: replaced

With any OpenShift deployment, we only install the cluster-level part
and allow infrastructure owner to "enable" the addition of user monitoring.

### User Stories

#### US1

As a service owner, I’d like to configure a ServiceMonitor CRD so that Red Hat’s supported monitoring solution is able to scrape metrics for my service.

#### US2

As a service owner, I’d like to configure alerting rules for my service so that Red Hat’s supported monitoring solution notifies me if problems turn up.

#### US3

> As a member of the operations team, I want to configure Prometheus one time so that all existing and any new namespaces will have access.

Reviewer 5: namespaces has access? access to what? (:

Reviewer 3: "access to self-service monitoring" ?

Reviewer 4: Yeah. This use case is taking out right from a customers mouth :D It's more about Prometheus scraping metrics ("access") from namespaces that already exist when you deployed Prometheus; as well as any new namespaces created by someone.

A bit confusing without much context indeed.

Author: added "access to user workload monitoring"

#### US4

As a member of the operations team, I'd like to add my own exporters so that I can monitor infrastructure services that are not monitored out-of-the-box.

> Note: This covers adding ServiceMonitor and PrometheusRule CRDs.

Reviewer 3: Not sure this note is needed here, but not against it

#### US5

As a developer working on a Red Hat product not part of the infrastructure, I’d like to integrate my metrics into a single, supported Prometheus stack so that they are available for my customers.

#### US6

As a service owner, I’d like to query the metrics I have access to from the OpenShift console.

#### US7

> As a member of the operations team, I’d like to make sure that one team doesn't have access to metrics from another team.

Reviewer 5: Within one cluster? Within one Prometheus sets, right?

Reviewer 2: i imagine as a developer/user i'd also like to make sure other teams/tenants don't have access to my metrics.
(same use case, different stakeholder)

Reviewer 3: I think replacing team with tenant here sets the expectations appropriately. How teams are distributed in an OpenShift cluster is subjective to each customer.

Author: replaced team with tenant.

#### US8

As a member of the operations team, I'd like the ability for application developers to access container cpu and memory usage metrics.

### Implementation Details/Notes/Constraints

#### Isolation

> The cluster monitoring as well as the user monitoring stack are composed of custom resources

Reviewer 5: ```suggestion
The cluster monitoring, as well as the user monitoring stack, are composed of the custom resources
```

reconciled by two separate prometheus operators.

> Both prometheus operators are under supervision of cluster monitoring operator.

Reviewer 5: ```suggestion
Both prometheus operators are under the supervision of cluster monitoring operator.
```

> This allows a seperation of concerns in terms of ugpradability, security settings, and scalability.

Reviewer 6: s/seperation/separation/

Reviewer 5: Also
```suggestion
This allows a separation of concerns in terms of upgradability, security settings, and scalability.
```

Author: fixed

The cluster monitoring Prometheus Operator continues to be deployed in the `openshift-monitoring` namespace.
The user monitoring Prometheus Operator will be deployed in the `openshift-user-monitoring` namespace.

> Currently, cluster monitoring operator scans all namespaces filtering openshift namespaces labelled with

Reviewer 5: ```suggestion
Currently, cluster monitoring operator scans all namespaces filtering `openshift` namespaces labeled with
```

`openshift.io/cluster-monitoring: "true"` and takes this set of namespaces as the _allowed list_ of reconcilable
namespaces for the existing cluster monitoring prometheus operator.

The same set of openshift namespaces that exist today will be configured as a _deny list_ of reconcilable namespaces
at the user monitoring prometheus operator.
Deny listing of namespaces was added recently in prometheus operator, see [3].

To make sure that users cannot add Prometheus instances in user namespaces,

> a new WIP feature [4] will ensure that reconcilation of Prometheus custom resources

Reviewer 5: ```suggestion
a new WIP feature [4] will ensure that reconciliation of Prometheus custom resources
```

will be limited to the `openshift-user-monitoring` namespace for the user monitoring prometheus operator
and `openshift-monitoring` namespace for the cluster monitoring prometheus operator.

![](user-namespaces-custom-resources.png)

User monitoring Prometheus Operator (green) then will reconcile Prometheus custom resources
in the `openshift-user-monitoring` namespace only.
It ignores Prometheus custom resources in any other namespace.
The same holds true for the cluster monitoring prometheus operator (red).

#### Aggregation

The front facing Kubernetes service has to make it as simple as possible
for integrators to leverage both user monitoring as well as cluster monitoring.

This is achieved by using Thanos Querier which allows to aggregate
and optionally deduplicate the cluster and user monitoring metrics backends
under single Prometheus Query endpoint.

Usage of the Thanos querier component inside cluster monitoring has many advantages:

> 1. Run-time deduplication of metrics available today.

Reviewer 5: What about aggregating 2 sources together (cluster + user monitoring Prometheus instances)? (Global View)

Author: added :+1:

> 2. Potentially gap-filling of metrics series in the Prometheus HA user and cluster monitoring setup in the future.

Reviewer 5: Isn't that the same as point 1? 

Also "in future" relies to cluster mon setup which is 1 replica, but it might sound like everything is in the future (:

Author: it is the same technically, but from a use case perspective, i think we can leave them separate (it was at least an eye-opener for me).

Removed "potentially" and "in the future", indeed, it is available today :+1:

3. Potentially add the possibility of adding a query cache for increased performance in the future.

See [5] and [6] for more details.

### Tenancy

![](user-monitoring-request.png)

Tenancy is achieved by leveraging the existing topology that is protecting cluster monitoring Prometheus already today.

> The kube-rbac-proxy sidecar will be deployed along with prom-label-proxy in the Thanos Querier Deployment.

Reviewer 3: I think it's worth linking out to the prom-label-proxy readme here as it describes the tenancy model rather well I think: https://github.com/openshift/prom-label-proxy

Author: added the reference to prom-label-proxy

### Risks and Mitigations

> Risks are mainly in the area of security, namely in the realm of accessing metrics a user has no permissions for.

Reviewer 5: Well I would say risks are also:

* QoD (we mitigate some of it by timeouts, but the saturation of services can be done) 
* Prometheus unavailability by ingestion due to aggressive scrape target with high cardinality

Author: spelled out QoD = Query of Death, and added those bullet points too :+1:

This is achieved by reusing the existing tenancy stack as well as with the newly added features for allow-lists,
deny-lists and reconcilation limiting settings for prometheus operator.

## Design Details

### Test Plan

TBD

### Graduation Criteria

N/A

#### Examples

These are generalized examples to consider, in addition to the aforementioned

> [maturity levels][maturity-levels].

Reviewer: Was this meant as a link or a ref to something?

Author: this is from the enhancement template :man_shrugging:

##### Dev Preview -> Tech Preview

There is no dev preview planned, just tech preview.

##### Tech Preview -> GA 

> - More testing (upgrade, downgrade, scale)

Reviewer: I believe we said we wanted to do one more thing for GA and that is the documentation of OLM prom-operator IIRC and removal of `OperatorGroup`?

Reviewer: Also I think we wanted to do some HA work for this, but not sure thats a requirement to go GA?

Author: agreed to both, added bullet points.

- Sufficient time for feedback
- Available by default

##### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

### Upgrade / Downgrade Strategy

Upgrades and downgrades will be orchestrated by cluster monitoring operator

> the same way as the cluster monitoring stack

Reviewer: I think something is missing here?

Author: I was in the german habit of writing a way too long sentence ;-) shortened and broke into two.

as user monitoring leverages the same semantics in terms of deployable assets
and custom resources.

### Version Skew Strategy

N/A

## Implementation History

N/A

## Drawbacks

N/A

## Alternatives

> N/A

Reviewer 3: Sticking to current OLM based prom operator that our customers are not happy with? (I don't think that's actually an option as we're literally doing this because no one including our own SRE team is happy with the current solution, but maybe it's worth nothing that here).

Author: indeed, added with a declared tendency that this is not a proper replacement, plus we would need to keep OperatorGroups.

## References

[1] https://github.com/brancz/kube-rbac-proxy

[2] https://github.com/openshift/prom-label-proxy

[3] https://github.com/coreos/prometheus-operator/pull/2710

[4] https://github.com/coreos/prometheus-operator/pull/2197

> [5] https://github.com/thanos-io/thanos/pull/1541

Reviewer 5: ```suggestion
[5] https://deploy-preview-1541--thanos-io.netlify.com/components/query.md/
```
For now (:
